### PR TITLE
Make shape a tuple

### PR DESCRIPTION
**Pull request by awni**
## Proposed changes

Switch to tuple shape in Python.

Companion PR in mlx-examples: https://github.com/ml-explore/mlx-examples/pull/393

### CODE REVIEW

**Comment by jagrit06** [comment]
@awni Do we need to go and change the docs for ops that require shape as an argument to be tuples ?
I don't think it's needed at the moment, but thought I should ask

**Comment by awni** [comment]
> Do we need to go and change the docs for ops that require shape as an argument to be tuples

Luckily I don't think so, for all of those ops the shape can be either a `tuple` or a `list`. That was already the case before this change. We may want to make it more clear in the docs but I'll leave that as a future update:

```
>>> mx.zeros([2, 2])
array([[0, 0],
       [0, 0]], dtype=float32)
>>> mx.zeros((2, 2))
array([[0, 0],
       [0, 0]], dtype=float32)
```

File: python/mlx/nn/losses.py [hunk 1]
```diff
@@ -70,8 +70,9 @@ def cross_entropy(
     targets_as_probs = targets.ndim == logits.ndim
 
     def _drop_dim(shape, axis):
+        shape = list(shape)
```
Comment (angeloskath) [inline python/mlx/nn/losses.py:73]: Feel free to ignore but `return shape[:axis] + shape[axis+1:]`? Avoids the list/tuple conversion.

**Review by angeloskath — APPROVED** [review]
Perfect!

**Review by jagrit06 — APPROVED** [review]
LGTM

File: python/mlx/nn/losses.py (hunk 1 above)
Comment (awni) [reply python/mlx/nn/losses.py:73]: Ah it's unfortunately a slightly more involved to get negative axes to work.

Comment (angeloskath) [reply python/mlx/nn/losses.py:73]: Makes sense :-)

**Comment by awni** [comment]
Ok I'm merging this!! Something is almost certainly going to break so just keep an eye out! I checked most things in core and examples though so that should be ok.

**Comment by dastrobu** [comment]
@awni this seems to close #297. 
It would be intriguing to learn the decisive factor that prompted the implementation of proposal #297.
